src/Utility/HttpResponsePaginator.php

```diff
@@ -0,0 +1,284 @@
+            if (! mb_strstr($params, 'page=')) {
+                $lastBuilder->withParam('page', 0);
+            }
+
+            // Is ?per_page present? If not, set a sane default.
+            if (! mb_strstr($params, 'per_page=')) {
+                $lastBuilder->withParam('per_page', 100);
+            }
+
+            // Ensure ?include_totals=true is present, required to iterate using this class.
+            $lastBuilder->withParam('include_totals', true);
+
+            // Increment our network request tracker for reference.
+            ++$this->requestCount;
+
+            // Issue next paged request.
+            try {
+                $lastBuilder->call();
+            } catch (NetworkException $exception) {
+                return false;
+            }
+
+            // Process the response.
+            return $this->processLastResponse();
+        }
+
+        return false;
+    }
+
+    /**
+     * Make a network request for the next page of results.
+     */
+    private function getNextResults(): bool
+    {
+        if ($this->lastBuilder() && $this->lastResponse()) {
+            // Retrieve the active HttpRequest instance to repeat the request.
+            $lastBuilder = $this->lastBuilder();
+
+            // Get the next page.
+            $page = ceil($this->position / $this->requestLimit);
+
+            // Set the next page.
+            $lastBuilder->withParam('page', $page);
+
+            // Increment our network request tracker for reference.
+            ++$this->requestCount;
+
+            // Issue next paged request.
+            try {
+                $lastBuilder->call();
+            } catch (NetworkException $exception) {
+                return false;
+            }
+
+            // Process the response.
+            return $this->processLastResponse();
+        }
+
+        return false;
+    }
+
+    /**
+     * Process the previous HttpResponse results and cache them for iterator content.
+     */
+    private function processLastResponse(): bool
+    {
+        $lastRequest = $this->lastRequest();
+        $lastResponse = $this->lastResponse();
+
+        if ($lastRequest && $lastResponse) {
+            // We can only paginate GET requests.
+            if (mb_strtolower($lastRequest->getMethod()) !== 'get') {
+                throw PaginatorException::httpMethodUnsupported();
+            }
+
+            // Was the HTTP request successful?
+            if (HttpResponse::wasSuccessful($lastResponse)) {
+                // Decode the response.
+                $results = HttpResponse::decodeContent($lastResponse);
+
+                // No results, abort processing.
+                if (! is_array($results) || ! count($results)) {
+                    return false;
+                }
+
+                // There is no 'start' key, the request was probably made without the include_totals param. Try again using safe pagination defaults.
+                if (! isset($results['start'])) {
```

Comment:
Is this retrying the request until it gets a response with a `start` field? Is there any danger of creating an infinite loop - if perhaps one of the endpoints unexpectedly didn't have a `start` field in its response